lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -265,39 +269,41 @@ class PolicyElementAssociation < ::ActiveRecord::Base
 
       belongs_to :user_attribute
       belongs_to :object_attribute
+      belongs_to :operation_set
 
       #TODO: ActiveRecord's generated operations= method is inefficient, makes 1 query for each op added or removed even though there's no hooks
       # Awkward manual implementation for now, but in the future change this to an hstore or something in the postgres adapter,
       # and/or fix Rails.
-      def operations=(updated_operations)
-        updated_operation_set = Set.new(updated_operations)
-        current_operation_set = Set.new(self.operations)
-        new_operations = updated_operation_set - current_operation_set
-        removed_operations = current_operation_set - updated_operation_set
+      def operations=(updated_operations_objects)
+        updated_set_of_operation_objects = Set.new(updated_operations_objects)
+        current_set_of_operation_objects = Set.new(self.operations)
+        new_operation_objects = updated_set_of_operation_objects - current_set_of_operation_objects
+        removed_operation_objects = current_set_of_operation_objects - updated_set_of_operation_objects
         transaction do
           OperationsPolicyElementAssociation.where(policy_element_association_id: self.id)
-                                            .where(operation_id: removed_operations.map(&:id))
+                                            .where(operation_id: removed_operation_objects.map(&:id))
                                             .delete_all
           OperationsPolicyElementAssociation.import([:policy_element_association_id, :operation_id],
-                                                     new_operations.map{ |op| [self.id, op.id] },
+                                                     new_operation_objects.map { |op| [self.id, op.id] },
                                                      validate: false)
         end
         self.clear_association_cache
       end
 
-      def self.add_association(user_attribute, operation_set, object_attribute, policy_machine_uuid)
+      def self.add_association(user_attribute, set_of_operation_objects, operation_set, object_attribute, policy_machine_uuid)
         where(
-          user_attribute_id: user_attribute.id,
-          object_attribute_id: object_attribute.id
-        ).first_or_create.operations = operation_set.to_a
+          user_attribute: user_attribute.id,
+          object_attribute_id: object_attribute.id,
+          operation_set_id: operation_set.id
+        ).first_or_create.operations = set_of_operation_objects.to_a
```

Comment:
We discussed this one and determined that we didn't want to change it for now because `import` uses an `on_duplicate_key_update` param. We'd been assuming associations were unique, but they aren't so importing actually creates another association even when the `user_attribute_id` and `object_attribute_id` are the same. That's undesirable. I think we may have also been planning on changing it once we added that uniqueness key.